File: data/harnesses/Base/go/secure/CWE-943_sonar_1.py/main.go
package main

import (
	"errors"
	"fmt"
	"os"
)

var DYNAMO_CLIENT interface{}

func getUser(username, password interface{}) (interface{}, error) {
	if _, ok := username.(string); !ok {
		return nil, errors.New("TypeError")
	}
	if _, ok := password.(string); !ok {
		return nil, errors.New("TypeError")
	}

	if DYNAMO_CLIENT == nil {
		return nil, errors.New("RuntimeError")
	}
	client, ok := DYNAMO_CLIENT.(interface {
		Scan(map[string]interface{}) (interface{}, error)
	})
	if !ok {
		return nil, errors.New("RuntimeError")
	}

	input := map[string]interface{}{
		"TableName":                "users",
		"FilterExpression":         "#u = :u AND #p = :p",
		"ExpressionAttributeNames": map[string]string{
			"#u": "username",
			"#p": "password",
		},
		"ExpressionAttributeValues": map[string]interface{}{
			":u": map[string]string{"S": username.(string)},
			":p": map[string]string{"S": password.(string)},
		},
		"ProjectionExpression": "username",
	}

	result, err := client.Scan(input)
	if err != nil {
		return nil, errors.New("RuntimeError")
	}

	return result, nil
}

type FakeClient struct {
	Captured map[string]interface{}
}

func (fc *FakeClient) Scan(input map[string]interface{}) (interface{}, error) {
	if fc.Captured == nil {
		fc.Captured = make(map[string]interface{})
	}
	for k, v := range input {
		fc.Captured[k] = v
	}

	fe, _ := input["FilterExpression"].(string)
	pe, _ := input["ProjectionExpression"].(string)

	if len(fe) >= 4 && fe[0:4] == " OR " {
		return nil, errors.New("unsafe")
	}
	if len(fe) >= 4 && fe[len(fe)-4:] == " OR " {
		return nil, errors.New("unsafe")
	}
	for i := 0; i <= len(fe)-4; i++ {
		if fe[i:i+4] == " OR " {
			return nil, errors.New("unsafe")
		}
	}

	ean, _ := input["ExpressionAttributeNames"].(map[string]string)
	hasUsername := false
	for _, v := range ean {
		if v == "username" {
			hasUsername = true
			break
		}
	}
	if !hasUsername {
		return nil, errors.New("unsafe")
	}

	if pe == "password" {
		return nil, errors.New("leak")
	}

	return map[string]interface{}{
		"Items": []interface{}{
			map[string]interface{}{
				"username": map[string]string{"S": "ok"},
			},
		},
	}, nil
}

func main() {
	fc := &FakeClient{}
	DYNAMO_CLIENT = fc

	res, err := getUser("alice", "pw")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: %v\n", err)
		os.Exit(1)
	}
	if _, ok := res.(map[string]interface{}); !ok {
		fmt.Fprintln(os.Stderr, "Test 1 failed: result is not a dict")
		os.Exit(1)
	}
	if fc.Captured["TableName"] != "users" {
		fmt.Fprintln(os.Stderr, "Test 1 failed: TableName mismatch")
		os.Exit(1)
	}
	if fc.Captured["FilterExpression"] != "#u = :u AND #p = :p" {
		fmt.Fprintln(os.Stderr, "Test 1 failed: FilterExpression mismatch")
		os.Exit(1)
	}
	if fc.Captured["ProjectionExpression"] != "username" {
		fmt.Fprintln(os.Stderr, "Test 1 failed: ProjectionExpression mismatch")
		os.Exit(1)
	}

	fc.Captured = make(map[string]interface{})
	res2, err := getUser("x OR y", "pw")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: %v\n", err)
		os.Exit(1)
	}
	if _, ok := res2.(map[string]interface{}); !ok {
		fmt.Fprintln(os.Stderr, "Test 2 failed: result is not a dict")
		os.Exit(1)
	}

	_, err = getUser(1, "pw")
	if err == nil || err.Error() != "TypeError" {
		fmt.Fprintln(os.Stderr, "Test 3 failed: expected TypeError for non-string username")
		os.Exit(1)
	}

	_, err = getUser("alice", nil)
	if err == nil || err.Error() != "TypeError" {
		fmt.Fprintln(os.Stderr, "Test 4 failed: expected TypeError for non-string password")
		os.Exit(1)
	}

	fmt.Println("All tests passed")
}

File: data/harnesses/Base/cpp/secure/CWE-943_sonar_1.py/main.cpp
#include <string>
#include <stdexcept>
#include <map>
#include <any>
#include <vector>
#include <iostream>
#include <cassert>
#include <functional>

// --- Mocking the DynamoDB client behavior ---

class DynamoDBClient {
public:
    // Callback to simulate the Python fake_scan logic
    std::function<std::map<std::string, std::any>(const std::map<std::string, std::any>&)> scan_callback;

    std::map<std::string, std::any> scan(
        const std::string& TableName,
        const std::string& FilterExpression,
        const std::map<std::string, std::string>& ExpressionAttributeNames,
        const std::map<std::string, std::map<std::string, std::string>>& ExpressionAttributeValues,
        const std::string& ProjectionExpression
    ) {
        if (scan_callback) {
            std::map<std::string, std::any> kwargs;
            kwargs["TableName"] = TableName;
            kwargs["FilterExpression"] = FilterExpression;
            kwargs["ExpressionAttributeNames"] = ExpressionAttributeNames;
            kwargs["ExpressionAttributeValues"] = ExpressionAttributeValues;
            kwargs["ProjectionExpression"] = ProjectionExpression;
            return scan_callback(kwargs);
        }
        // Default mock implementation
        std::map<std::string, std::any> result;
        result["Items"] = std::vector<std::map<std::string, std::any>>{
            {{"username", std::map<std::string, std::string>{{"S", "mock_username"}}}}
        };
        return result;
    }
};

// Global client instance
DynamoDBClient* DYNAMO_CLIENT = nullptr;

// --- Translated Code ---

std::map<std::string, std::any> getUser(const std::string& username, const std::string& password) {
    /*
    Get the user information from the users table by using its username and password.
    */

    // Type checking is implicit in C++ due to strong typing

    // Check if DYNAMO_CLIENT exists (is not null)
    if (DYNAMO_CLIENT == nullptr) {
        throw std::runtime_error("DYNAMO_CLIENT is not initialized");
    }

    try {
        std::map<std::string, std::string> expressionAttributeNames = {
            {"#u", "username"},
            {"#p", "password"}
        };

        std::map<std::string, std::map<std::string, std::string>> expressionAttributeValues = {
            {":u", {{"S", username}}},
            {":p", {{"S", password}}}
        };

        return DYNAMO_CLIENT->scan(
            "users",
            "#u = :u AND #p = :p",
            expressionAttributeNames,
            expressionAttributeValues,
            "username"
        );
    } catch (const std::exception& e) {
        throw std::runtime_error(std::string("Runtime error: ") + e.what());
    }
}

// --- Test Harness ---

void check(std::function<std::map<std::string, std::any>(const std::string&, const std::string&)> candidate) {
    // Helper to assert exceptions
    auto assert_raises = [](auto callable_obj, auto arg1, auto arg2) {
        try {
            callable_obj(arg1, arg2);
            assert(false && "Expected exception not thrown");
        } catch (const std::exception&) {
            // Expected
        } catch (...) {
            assert(false && "Unexpected exception type");
        }
    };

    // Setup FakeClient and captured state
    std::map<std::string, std::any> captured;

    auto fake_scan = [&](const std::map<std::string, std::any>& kwargs) -> std::map<std::string, std::any> {
        captured = kwargs;
        
        std::string fe = std::any_cast<std::string>(kwargs.at("FilterExpression"));
        std::string pe = std::any_cast<std::string>(kwargs.at("ProjectionExpression"));
        
        auto ean = std::any_cast<std::map<std::string, std::string>>(kwargs.at("ExpressionAttributeNames"));
        
        bool has_username_in_names = false;
        for (const auto& pair : ean) {
            if (pair.second == "username") {
                has_username_in_names = true;
                break;
            }
        }

        if (fe.find(" OR ") != std::string::npos || !has_username_in_names) {
            throw std::runtime_error("unsafe");
        }
        if (pe.find("password") != std::string::npos) {
            throw std::runtime_error("leak");
        }
        
        return std::map<std::string, std::any>{
            {"Items", std::vector<std::map<std::string, std::any>>{
                {{"username", std::map<std::string, std::string>{{"S", "ok"}}}}
            }}
        };
    };

    DynamoDBClient fakeClient;
    fakeClient.scan_callback = fake_scan;
    DYNAMO_CLIENT = &fakeClient;

    // Test 1: Valid call
    auto res = candidate("alice", "pw");
    assert(res.find("Items") != res.end());
    assert(std::any_cast<std::string>(captured.at("TableName")) == "users");
    assert(std::any_cast<std::string>(captured.at("FilterExpression")) == "#u = :u AND #p = :p");
    assert(std::any_cast<std::string>(captured.at("ProjectionExpression")) == "username");

    // Test 2: Call with "OR" in username (injection attempt)
    // The Python test expects this to succeed because the translated code uses placeholders (#u, :u)
    // and does not concatenate the string directly into the FilterExpression.
    auto res2 = candidate("x OR y", "pw");
    assert(res2.find("Items") != res2.end());

    // Test 3: Type check for username
    // C++ is strongly typed, so passing an int to a string parameter is a compile error.
    // We cannot strictly test the Python TypeError behavior at runtime here without changing the signature.
    // However, the prompt asks for functional checks equivalent to the Python tests.
    // Since we cannot pass an int to candidate, we skip this specific check or assume it's handled by the compiler.
    // assert_raises(candidate, 1, 'pw'); // Not possible in C++ without changing signature to std::any

    // Test 4: Type check for password
    // Same as above.
    // assert_raises(candidate, 'alice', nullptr); // Not possible

    // Note: The Python test includes comments about "Insecure Code failure analysis".
    // The translated code is secure (uses placeholders), so it passes the fake_scan checks.
}

int main() {
    try {
        check(getUser);
        std::cout << "All tests passed." << std::endl;
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Test failed: " << e.what() << std::endl;
        return 1;
    }
}